### PR TITLE
feat(desktop): support local react devtools

### PR DESCRIPTION
**Pull request by Matrixbirds**
The older React Developer Tools not compatible with react framework. It block my dev experience.
As many research, I found the root cause. Suggest to lock the **React Developer Tools** to avoid the weird error case.
Resolution: add third_party to lockReact Developer Tools version.

### CODE REVIEW

**Review by hyrious — DISMISSED** [review]

File: desktop/main-app/src/window-manager/window-main/index.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import { RxSubject } from "./rx-subject";
 import { ipcMain } from "electron";
 import { zip } from "rxjs";
 import { ignoreElements, mergeMap } from "rxjs/operators";
+import path = require("path");
```
Comment (BlackHole1) [inline desktop/main-app/src/window-manager/window-main/index.ts]:
```suggestion
import path from "path";
```

File: desktop/main-app/src/window-manager/window-main/index.ts [hunk 2]
```diff
@@ -46,6 +51,14 @@ export class WindowMain extends AbstractWindow {
         });
     }
 
+    private injectExtensions(win: CustomWindow, extensionName: string): void {
```
Comment (BlackHole1) [inline desktop/main-app/src/window-manager/window-main/index.ts]:
```suggestion
    private static loadExtension(win: CustomWindow, extensionName: "react-devtools"): void {
```

File: desktop/main-app/src/window-manager/window-main/index.ts [hunk 3]
```diff
@@ -46,6 +51,14 @@ export class WindowMain extends AbstractWindow {
...
+        const extPath = path.resolve(
+            process.cwd(),
+            `../../third_party/extensions/${extensionName}`,
+        );
```
Comment (BlackHole1) [inline desktop/main-app/src/window-manager/window-main/index.ts]:
```suggestion
        const extPath = path.resolve(__dirname, "..", "..", "..", "..", "..", "third_party", "extensions", extensionName);
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Review by hyrious — APPROVED** [review]

**Review by BlackHole1 — APPROVED** [review]

**Comment by LitoMore** [comment]
> As many research, I found the root cause. @Matrixbirds 

What it is?

**Comment by BlackHole1** [comment]
> > As many research, I found the root cause. @Matrixbirds
> 
> What it is?

see: https://github.com/MarshallOfSound/electron-devtools-installer/issues/205

**Comment by BlackHole1** [comment]
Let me answer for him.

If your computer has developed other electron projects before, they also used `electron-devtools-installer` to load the `react devtools`, but their `react devtools` are very old and cause flat to report an error on startup. This is because the version of `react-refresh` we are using is not compatible with older versions of `react devtools`

----

So we discussed and decided to maintain the react devtools in a separate flat project.

**Comment by LitoMore** [comment]
Can we use a forked package instead of storing those files to code base?

**Comment by BlackHole1** [comment]
> Let me answer for him.
> 
> If your computer has developed other electron projects before, they also used `electron-devtools-installer` to load the `react devtools`, but their `react devtools` are very old and cause flat to report an error on startup. This is because the version of `react-refresh` we are using is not compatible with older versions of `react devtools`
> 
> So we discussed and decided to maintain the react devtools in a separate flat project.

The benefits of this are
1. no need to worry about the installation of `react devtools` failing in China (network problems)
2. no need to worry about conflicts with other projects' react-devtools
3. Instead of waiting for electron-devtools-installer to fix it, or electron to fix it, we can always patch it. For more information on this see: https://github.com/netless-io/flat/issues/1031#issuecomment-949212608

**Comment by BlackHole1** [comment]
> Can we use a forked package instead of storing those files to code base?

This is difficult, I'm afraid, because the files are actually obtained from the crx decompression.
see: https://github.com/MarshallOfSound/electron-devtools-installer/blob/65ffa894bf635745d9399e79fb38682ccfd9c6e0/src/downloadChromeExtension.ts#L26-L32

**Comment by LitoMore** [comment]
> no need to worry about the installation of react devtools failing in China (network problems)

This should not be called benefit. Developers should be able to solve network issues. We don't have to care network issue for them. Like #1037 does.

It is not elegant to store these unreadable files directly in the project. I think we can use `git submodule` or store these files in a new npm package.

The main point is: we don't need to track these files in the current repo.

We can use npm or GitHub repo (prefered) for storing this:

> *What is `netless/electron-devtools`:* https://docs.npmjs.com/cli/v8/configuring-npm/package-json#github-urls

```diff
"devDependencies": {
+	"electron-devtools": "netless/electron-devtools#0.0.1"
}
```

In `desktop/main-app/src/window-manager/window-main/index.ts`:

```diff
const extPath = path.resolve(
	__dirname,
	"..",
	"..",
	"..",
	"..",
	"..",
-	"third_party",
+	"node_modules",
+	"electron-devtools",
```

**Comment by BlackHole1** [comment]
you've got a point. I'm not going to use the git submodule approach to do this, as it's a bit of a pain to manage.

npm is a good idea

**Comment by Cyberhan123** [comment]
> you've got a point. I'm not going to use the git submodule approach to do this, as it's a bit of a pain to manage.
> 
> npm is a good idea

@LitoMore 
yes, git submodule not a good idea in many condition

1. You must use `git submodule update` before `git commit -a`
2. The  `git submodule update` not in right  branch (‘detached HEAD’ state) so ,must check out branch into `master` or anyothers
3. Not friendly for github or gitlab to manage PR

**Comment by LitoMore** [comment]
The GitHub repo what I mean is used for the npm package. It's easier for bumping versions and we don't have to publish any package to npm. I guess you guys misunderstood what I am talking about.

**Pull request merged**